### ISSUE TITLE
cut value starting with numerals interpreted as int even for column with type 'text'

### ISSUE THREAD

**Issue by jbothma** [openspending/babbage#11]
I expect the query value 904 to match facts where the referred text column has the value 9
04. I also expect that the query value "904aaa" would be interpreted as that string and not "904".

e.g.

    curl  'localhost:5000/api/cubes/incexp/facts?cut=function.function_code:904.123aaa'

results in 

```
ProgrammingError: (psycopg2.ProgrammingError) operator does not exist: text = integer
LINE 4: WHERE incexp_labeled.function_code = 904) AS anon_1
                                           ^
HINT:  No operator matches the given name and argument type(s). You might need to add explicit type casts.
 [SQL: 'SELECT count(%(param_1)s) AS count_1 \nFROM (SELECT incexp_labeled.function_code AS "function.function_code", incexp_labeled.function_desc AS "function.function_desc", incexp_labeled.incexp_cde AS "incexp_code.incexp_code", incexp_labeled.description AS "incexp_code.incexp_desc", incexp_labeled.period AS "period.period", incexp_labeled.demarcation_code AS "demarcation_code.demarcation_code", incexp_labeled.act_or_bud_amt AS amount \nFROM incexp_labeled \nWHERE incexp_labeled.function_code = %(param_2)s) AS anon_1'] [parameters: {'param_1': True, 'param_2': 904}]
127.0.0.1 - - [04/Apr/2016 13:17:25] "GET /api/cubes/incexp/facts?cut=function.function_code:904.123aaa HTTP/1.1" 500 -
```

same with cut value of just 900

```
ProgrammingError: (psycopg2.ProgrammingError) operator does not exist: text = integer
LINE 4: WHERE incexp_labeled.function_code = 904) AS anon_1
                                           ^
HINT:  No operator matches the given name and argument type(s). You might need to add explicit type casts.
 [SQL: 'SELECT count(%(param_1)s) AS count_1 \nFROM (SELECT incexp_labeled.function_code AS "function.function_code", incexp_labeled.function_desc AS "function.function_desc", incexp_labeled.incexp_cde AS "incexp_code.incexp_code", incexp_labeled.description AS "incexp_code.incexp_desc", incexp_labeled.period AS "period.period", incexp_labeled.demarcation_code AS "demarcation_code.demarcation_code", incexp_labeled.act_or_bud_amt AS amount \nFROM incexp_labeled \nWHERE incexp_labeled.function_code = %(param_2)s) AS anon_1'] [parameters: {'param_1': True, 'param_2': 904}]
127.0.0.1 - - [04/Apr/2016 13:19:18] "GET /api/cubes/incexp/facts?cut=function.function_code:904 HTTP/1.1" 500 -
```

**Comment by akariv** [comment]
You can try and wrap these values with double quotes, e.g.

`curl  'localhost:5000/api/cubes/incexp/facts?cut=function.function_code:"904.123aaa"'`

**Comment by jbothma** [comment]
That seems to do the trick. Thanks

**Comment by jbothma** [comment]
Why does babbage add the possibility of quoted/escaped strings over the cubes API?

Is coercing values to column type left out intentionally or have you just not gotten to it yet?

I have some text columns with integer values that I can't query since babbage parses them as ints and then can't compare the int to the string in the DB. I'm trying to think of the best way to solve this. Perhaps simple type coercion based on column type, and fixing the parser interpreting a string value starting with an int as an int is easier than having clients know when to quote values and when not to.

**Comment by jbothma** [comment]
Ah I see there is a call to sqlalchemy.type_coerce but from the docs it looks like it coerces the value of a result python-side, not of a query restriction value?

**Comment by jbothma** [comment]
Hey there. What do you think of https://github.com/Code4SA/babbage/pull/7 ?

We need this because we have some attributes that are really strings, and the API will parse a value like 0200 as the int 200 and not match in the DB if the value gets cast/coerced for the DB query. It turns out instead of hassling more with the parser we can just simplify it and require strings to be quoted. I'm not sure everything's in the right place yet and I cheated a bit by not adding type info to measures at least for the sake of being consistent across subclasses of Concept, but it solves our immediate problem and I'd like feedback from you before putting more effort into it. Also I'd have to rebase on your master before bringing PRing to this repo. Your comments are very welcome here or in https://github.com/Code4SA/babbage

**Comment by akariv** [comment]
Hi,

Apart from a few style issues and the fact that the change is not backwards compatible, I'm not sure I understand the need for it.
The current implementation allows you to provide type-exact parameters (by quoting strings), which to me seems like good practice for anyone using the API. This is why it feels wrong to me to introduce a breaking change to add functionality that is already possible with the current implementation.

**Comment by jbothma** [comment]
I agree on the style issues and not breaking backward compatibility.

The type_coerce wasn't doing what was expected of it resulting in errors from the DB when a cut on a string column was parsed as an integer. When I changed the type_coerce to cast, it worked as intended (a cut value parsed as an integer could be used on a string column) but I can now get a successful response with 0 values as a consequence of an unintended cast - the unquoted cut value 0200 would be parsed as the int 200, and the databse query would cast it to the string "200" which won't match the rows with value "0200" that I was looking for.

The issue is that the user doesn't get feedback about magic done on their behalf and simply don't get the data they asked for. The change was intended to give a helpful error instead of a false negative.

My personal preference is to just simplify the API and remove the unquoted strings and we can still break backward compatibility so I thought I'd see if that's what you want too. Now I believe we want unquoted strings as long as the user gets an error response indicating that their argument was parsed as something that doesn't make sense to compare against the concept they're cutting on. So I'll revert the parser but still remove casting and tidy up the type checking.

Another way to solve the specific issue we've identified is to change the parser to parse anything starting with 0 as a string. That might fix everything but I didn't want to sit and prove that that's the only kind of integer that can be misinterpreted. It might also surprise some users that 0200 gets parsed as a string and this change explains everything clearly.
